crates/bevy_remote/src/http.rs

```diff
@@ -0,0 +1,246 @@
+//! The BRP transport using JSON-RPC over HTTP.
+//!
+//! Adding the [`RemoteHttpPlugin`] to your [`App`] causes Bevy to accept
+//! connections over HTTP (by default, on port 15702) while your app is running.
+//!
+//! Clients are expected to `POST` JSON requests to the root URL; see the `client`
+//! example for a trivial example of use.
+
+#![cfg(not(target_family = "wasm"))]
+
+use crate::{error_codes, BrpBatch, BrpError, BrpMessage, BrpRequest, BrpResponse, BrpSender};
+use anyhow::Result as AnyhowResult;
+use async_channel::Sender;
+use async_io::Async;
+use bevy_app::{App, Plugin, Startup};
+use bevy_ecs::system::{Res, Resource};
+use bevy_tasks::IoTaskPool;
+use core::net::{IpAddr, Ipv4Addr};
+use http_body_util::{BodyExt as _, Full};
+use hyper::{
+    body::{Bytes, Incoming},
+    server::conn::http1,
+    service, Request, Response,
+};
+use serde_json::Value;
+use smol_hyper::rt::{FuturesIo, SmolTimer};
+use std::net::TcpListener;
+use std::net::TcpStream;
+
+/// The default port that Bevy will listen on.
+///
+/// This value was chosen randomly.
+pub const DEFAULT_PORT: u16 = 15702;
+
+/// The default host address that Bevy will use for its server.
+pub const DEFAULT_ADDR: IpAddr = IpAddr::V4(Ipv4Addr::new(127, 0, 0, 1));
+
+/// Add this plugin to your [`App`] to allow remote connections over HTTP to inspect and modify entities.
+/// It requires the [`RemotePlugin`](super::RemotePlugin).
+///
+/// This BRP transport cannot be used when targeting WASM.
+///
+/// The defaults are:
+/// - [`DEFAULT_ADDR`] : 127.0.0.1.
+/// - [`DEFAULT_PORT`] : 15702.
+pub struct RemoteHttpPlugin {
+    /// The address that Bevy will bind to.
+    address: IpAddr,
+    /// The port that Bevy will listen on.
+    port: u16,
+}
+
+impl Default for RemoteHttpPlugin {
+    fn default() -> Self {
+        Self {
+            address: DEFAULT_ADDR,
+            port: DEFAULT_PORT,
+        }
+    }
+}
+
+impl Plugin for RemoteHttpPlugin {
+    fn build(&self, app: &mut App) {
+        app.insert_resource(HostAddress(self.address))
```

Comment:
This overwrites any previous value. Should we insert only if the user didn't, to allow them to overwrite the default address and port?

I appreciate you can configure via the methods, but since world resources are "public", I think the more common pattern in Bevy is to let the user manually insert them before a plugin if they want to override any default. @alice-i-cecile any opinion on that?

Reply:
Yeah, I'd prefer a manual insertion with docs.